File: bc-w3/bcw3/Basic/List/List.cpp
#include <iostream>
#include <cstdlib>
#include "List.h"

List::List(int capacity, double multiplier)
    : capacity(capacity), multiplier(multiplier), current(0) {
        this->array = (int*)malloc(sizeof(int)*capacity);
        if ( this->array == NULL ) {
            throw OutOfMemoryException();
        }
}

List::~List() {
    free(array);
}

int List::size() const {
    return current;
}

int List::max_size() const {
    return capacity;
}

void List::erase(int index) {
    int last = this->current - 1;
    
    if ( this->current == 0 ) {
        throw ZeroLenException();
    }
    for ( int i = index; i < last; i++ ) {
        array[i] = array[i+1];
    }
    this->current -= 1;
}

void List::insert(int value, int index) {
    int newCurrent = this->current + 1;
    
    if ( newCurrent > this->capacity ) {
        int newCapacity = this->capacity * this->multiplier;
        int* newArray;
        
        if ( newCapacity == this->capacity ) {
            newCapacity += 5;
        }
        
        newArray = (int*)realloc(this->array, sizeof(int)*newCapacity);
        
        if ( newArray == NULL ) {
            throw OutOfMemoryException();
        }
        this->capacity = newCapacity;
        this->array = newArray;
    }
    
    for ( int i = newCurrent - 1; i > index; i-- ) {
        this->array[i] = this->array[i-1];
    }
    this->array[index] = value;
    this->current = newCurrent;
}

int List::find(int index) const {
    return this->array[index];
}

void List::push_back(int value) {
    this->insert(value, this->current);
}

int List::pop_back() {
    if ( this->current == 0 ) {
        throw ZeroLenException();
    }
    this->current -= 1;
    
    return this->array[this->current];
}

void List::sort() {
    int stack[this->current];
    int top = -1;
    
    if ( this->current < 2 ) {
        return;
    }
    
    top += 1;
    stack[top] = 0;
    top += 1;
    stack[top] = this->current - 1;
    
    while ( top >= 0 ) {
        int hi = stack[top--];
        int lo = stack[top--];
        int temp = this->array[hi];
        int i = lo - 1;
        int pivot;
        
        for ( int j = lo; j < hi; j++ ) {
            if ( this->array[j] <= temp ) {
                int buf = this->array[j];
                
                this->array[j] = this->array[++i];
                this->array[i] = buf;
            }
        }
        pivot = i + 1;
        this->array[hi] = this->array[pivot];
        this->array[pivot] = temp;
        
        if ( pivot - 1 > lo ) {
            top += 1;
            stack[top] = lo;
            top += 1;
            stack[top] = pivot - 1;
        }
        
        if ( pivot + 1 < hi ) {
            top += 1;
            stack[top] = pivot + 1;
            top += 1;
            stack[top] = hi;
        }
    }
}

int List::operator[](int index) const {
    return this->array[index];
}

bool List::operator==(const List& other) const {
    bool isEqual = true;
    int thisLast = this->current - 1;
    
    for ( int i = 0; i < other.current && isEqual; i++ ) {
        int otherValue = other.array[i];
        
        for ( int j = 0; j < this->current; j++ ) {
            if ( this->array[j] == otherValue ) {
                break;
            }
            if ( j == thisLast ) {
                isEqual = false;
            }
        }
    }
    
    return isEqual;
}

bool List::operator!=(const List& other) const {
    return !(*this == other);
}

std::ostream& operator<<(std::ostream& out, const List& list) {
    int last = list.size() - 1;
    
    if ( last < 0 ) {
        return out;
    }
    
    for ( int i = 0; i < last; i++ ) {
        out << list[i] << ' ';
    }
    return out << list[last];
}

File: bc-w3/java/Basic/List/List.java
import java.util.Arrays;

public class List {
    private int capacity;
    private double multiplier;
    private int current;
    private Integer[] array;
    
    public List(int capacity, double multiplier) throws OutOfMemoryException {
        this.capacity = capacity;
        this.multiplier = multiplier;
        this.current = 0;
        try {
            this.array = new Integer[capacity];   
        } catch (OutOfMemoryError e) {
            throw new OutOfMemoryException();
        }
    }
    
    public List(int capacity) throws OutOfMemoryException {
        this(capacity, 1.05);
    }
    
    public List() throws OutOfMemoryException {
        this(100, 1.05);
    }
    
    public int size() {
        return current;
    }
    
    public int max_size() {
        return capacity;
    }
    
    public void erase(int index) throws ZeroLenException, OutOfMemoryException {
        int numMoved = this.current - 1 - index;
        
        if ( this.current == 0 ) {
            throw new ZeroLenException();
        }
        
        if ( numMoved > 0 ) {
            try {
                System.arraycopy(this.array, index+1, this.array, index, numMoved);   
            } catch (OutOfMemoryError e) {
                throw new OutOfMemoryException();
            }
        }
        this.current -= 1;
        array[this.current] = null;
    }
    
    public void insert(int value, int index) throws OutOfMemoryException {
        int newCurrent = this.current + 1;
        
        try {
            if ( newCurrent > this.capacity ) {
                int newCapacity = (int) (this.capacity * this.multiplier);
                
                if ( newCapacity == this.capacity ) {
                    newCapacity += 5;
                }
                this.array = Arrays.copyOf(this.array, newCapacity);
                this.capacity = newCapacity;
            }
            System.arraycopy(this.array, index, this.array, index+1, this.current-index);
        } catch (OutOfMemoryError e) {
            throw new OutOfMemoryException();
        }
        this.array[index] = value;
        this.current += 1;
    }
    
    public int find(int index) {
        return this.array[index];
    }
    
    public void push_back(int value) throws OutOfMemoryException {
        this.insert(value, this.current);
    }
    
    public int pop_back() throws ZeroLenException {
        int eraseValue;
        
        if ( this.current == 0 ) {
            throw new ZeroLenException();
        }
        this.current -= 1;
        eraseValue = this.array[this.current];
        this.array[this.current] = null;
        
        return eraseValue;
    }
    
    public void sort() throws OutOfMemoryException {
        int[] stack;
        int top = -1;
        
        if ( this.current < 2 ) {
            return;
        }
        
        try {
            stack = new int[this.current];
        } catch (OutOfMemoryError e) {
            throw new OutOfMemoryException();
        }
        
        top += 1;
        stack[top] = 0;
        top += 1;
        stack[top] = this.current - 1;
        
        while ( top >= 0 ) {
            int hi = stack[top--];
            int lo = stack[top--];
            int temp = this.array[hi];
            int i = lo - 1;
            int pivot;
            
            for ( int j = lo; j < hi; j++ ) {
                if ( this.array[j] <= temp ) {
                    int buf = this.array[j];
                    
                    this.array[j] = this.array[++i];
                    this.array[i] = buf;
                }
            }
            pivot = i + 1;
            this.array[hi] = this.array[pivot];
            this.array[pivot] = temp;
            
            if ( pivot - 1 > lo ) {
                top += 1;
                stack[top] = lo;
                top += 1;
                stack[top] = pivot - 1;
            }
            
            if ( pivot + 1 < hi ) {
                top += 1;
                stack[top] = pivot + 1;
                top += 1;
                stack[top] = hi;
            }
        }
    }
    
    public boolean equals(List other) throws ZeroLenException, OutOfMemoryException {
        boolean isEqual = true;
        int thisLast = this.current - 1;
        Integer[] thisArrayCopy;
        
        if ( this.current != other.current ) {
            return false;
        }
        
        try {
            thisArrayCopy = this.array.clone();
        } catch (OutOfMemoryError e) {
            throw new OutOfMemoryException();
        }
        
        for ( int i = 0; i < other.current && isEqual; i++ ) {
            int otherValue = other.array[i];
            
            for ( int j = 0; j < this.current; j++ ) {
                if ( thisArrayCopy[j] != null && thisArrayCopy[j] == otherValue ) {
                    thisArrayCopy[j] = null;
                    break;
                } else if ( j == thisLast ) {
                    isEqual = false;
                }
            }
        }
        
        return isEqual;
    }
    
    public String toString() {
        int last = this.current - 1;
        StringBuffer out = new StringBuffer();
        
        if ( last < 0 ) {
            return "";
        }
        
        for ( int i = 0; i < last; i++ ) {
            out.append(this.array[i] + " ");
        }
        out.append(this.array[last]);
        
        return out.toString();
    }
}
